presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -277,29 +278,52 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, PlanVariableAllocat
 
         private Multimap<VariableReferenceExpression, VariableReferenceExpression> extractCorrelatedSymbolsMapping(List<Expression> correlatedConjuncts)
         {
-            // TODO: handle coercions and non-direct column references
             ImmutableMultimap.Builder<VariableReferenceExpression, VariableReferenceExpression> mapping = ImmutableMultimap.builder();
             for (Expression conjunct : correlatedConjuncts) {
                 if (!(conjunct instanceof ComparisonExpression)) {
                     continue;
                 }
 
                 ComparisonExpression comparison = (ComparisonExpression) conjunct;
-                if (!(comparison.getLeft() instanceof SymbolReference
-                        && comparison.getRight() instanceof SymbolReference
-                        && comparison.getOperator().equals(EQUAL))) {
+                // handle coercions and non-direct column references
+                Expression leftExpression;
+                Expression rightExpression;
+                if (comparison.getLeft() instanceof Cast) {
```

Comment:
won't it be a problem that we're getting rid of the casts if they are needed?

Reply:
not clear what you meant here. The logic is to extract both left and right sides of the comparsion expression if it is formed of CAST expressions in order to find the symbol mapping. If there's no CAST expression, then the logic goes BAU to check for symbolReference Type, or to skip to the next iteration.

Reply:
What I mean is that this result is then used as the filter for a join https://github.com/prestodb/presto/blob/0df58706580bc9803ab0a33354d47e25ae63b372/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedLateralJoinToJoin.java#L66.  Couldn't the missing cast be an issue?

Reply:
actually there is no changes to the correlated predicate in `DecorrelatedResult`. The logic is to extract the symbol of column reference from each side of predicate and create a direct variable mapping. As a decorrelated result, the original predicate still contains expression in Cast while the variable mapping relationship is created. The unit test validates this scenario.
![image](https://user-images.githubusercontent.com/32091981/107315043-c0d15300-6a4a-11eb-9513-ba19434885fb.png)

Reply:
That makes sense. thanks for the explanation.